internal/controller/resourceset_controller.go

```diff
@@ -294,38 +294,121 @@ func (r *ResourceSetReconciler) checkDependencies(ctx context.Context,
 	return nil
 }
 
+// getInputs fetches all the referenced provider objects and appends
+// the inputs from all of them. It returns the combined list of input
+// maps, including the static inputs specified on the ResourceSet.
+// Objects are deduplicated by their GVK, name and namespace.
+// The API version, kind, name and namespace of the provider are added
+// to each input map.
 func (r *ResourceSetReconciler) getInputs(ctx context.Context,
 	obj *fluxcdv1.ResourceSet) ([]map[string]any, error) {
-	providers := make([]fluxcdv1.InputProvider, 0)
-	providers = append(providers, obj)
-	for _, inputSource := range obj.Spec.InputsFrom {
-		var provider fluxcdv1.InputProvider
-		key := client.ObjectKey{
-			Namespace: obj.GetNamespace(),
-			Name:      inputSource.Name,
+
+	// inputProviderKey is the key used to identify input providers.
+	type inputProviderKey struct {
+		GVK       schema.GroupVersionKind
+		Name      string
+		Namespace string
+	}
+
+	// getInputProviderKey returns the key for the input provider.
+	getInputProviderKey := func(provider fluxcdv1.InputProvider) inputProviderKey {
+		return inputProviderKey{
+			GVK:       provider.GroupVersionKind(),
+			Name:      provider.GetName(),
+			Namespace: provider.GetNamespace(),
 		}
+	}
```

Comment:
I suggest moving the `InputProviderKey` type and all the logic around it to `internal/inputs`. The `Get` function could accept the controller-runtime `Object` so it will work for both RSET and RSIP.